plugins/catalog-backend-module-azure/src/providers/AzureDevOpsEntityProvider.ts

```diff
@@ -0,0 +1,167 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TaskRunner } from '@backstage/backend-tasks';
+import { Config } from '@backstage/config';
+import { AzureIntegration, ScmIntegrations } from '@backstage/integration';
+import {
+  EntityProvider,
+  EntityProviderConnection,
+  LocationSpec,
+  locationSpecToLocationEntity,
+} from '@backstage/plugin-catalog-backend';
+import { readAzureDevOpsConfigs } from './config';
+import { Logger } from 'winston';
+import { AzureDevOpsConfig } from './types';
+import * as uuid from 'uuid';
+import { codeSearch, CodeSearchResultItem } from '../lib';
+
+/**
+ * Provider which discovers catalog files within an Azure DevOps repositories.
+ *
+ * Use `AzureDevOpsEntityProvider.fromConfig(...)` to create instances.
+ *
+ * @public
+ */
+export class AzureDevOpsEntityProvider implements EntityProvider {
+  private readonly logger: Logger;
+  private readonly scheduleFn: () => Promise<void>;
+  private connection?: EntityProviderConnection;
+
+  static fromConfig(
+    configRoot: Config,
+    options: {
+      logger: Logger;
+      schedule: TaskRunner;
+    },
+  ): AzureDevOpsEntityProvider[] {
+    const providerConfigs = readAzureDevOpsConfigs(configRoot);
+
+    return providerConfigs.map(providerConfig => {
+      const integration = ScmIntegrations.fromConfig(configRoot).azure.byHost(
+        providerConfig.host,
+      );
+
+      if (!integration) {
+        throw new Error(
+          `There is no Azure integration for host ${providerConfig.host}. Please add a configuration entry for it under integrations.azure`,
+        );
+      }
+
+      return new AzureDevOpsEntityProvider(
+        providerConfig,
+        integration,
+        options.logger,
+        options.schedule,
+      );
+    });
+  }
+
+  private constructor(
+    private readonly config: AzureDevOpsConfig,
+    private readonly integration: AzureIntegration,
+    logger: Logger,
+    schedule: TaskRunner,
+  ) {
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+
+    this.scheduleFn = this.createScheduleFn(schedule);
+  }
+
+  private createScheduleFn(schedule: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return schedule.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.logger.child({
+            class: AzureDevOpsEntityProvider.prototype.constructor.name,
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.getProviderName} */
+  getProviderName(): string {
+    return `azureDevOps-provider:${this.config.id}`;
```

Comment:
we have a convention for these names? I see another entity provider named `MicrosoftGraphOrgEntityProvider:{...}`

So maybe it's better to go with `AzureDevOpsEntityProvider:{...}`? It'll differ from AWS S3 though...

Reply:
IMHO: I prefer AzureDevOpsEntityProvider as it matches the class name, easier to search for references.